Add label, assignee, milstone filter.

lanieerts opened the pull request:
Throttling은 실제 코드에 어떻게 적용 시켜야 할 지 이해가 안되서 빠졌습니다 orz
=> 수정. npm님은 모든것을 들고계십니다 ㄷㄷ Throttling 추가!

lanieerts committed:
Seperate reducer about comment and modify query.

lanieerts committed:
Modify detail of query for readMoreComments

lanieerts committed:
Modify url of action for createComment.

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

pjhjohn committed:
Rebase commit of upstream/master.
Introduce react-router-bootstrap to avoid page refresh

Removed legacy style for Tab. Removed navbar radius by overriding style

Resolved #32 : Implemented Sticky Footer

Reference : https://getbootstrap.com/examples/sticky-footer

Disabled default visibility of ReduxDevTools

CRLF -> LF

reference : http://stackoverflow.com/questions/1889559/git-diff-to-ignore-m

Comment, PostDetail, Post re-styled with react-bootstrap

* Enabled well class in _bootstrap.scss

UI Enhancement using Panel Component

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Merge branch 'master' of https://github.com/lanieerts/jsonplaceholder-client

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Add issues page on about page.

lanieerts committed:
Modify issues about #48.
1. Rename GithubIssues component.
2. Using ES6 templete literals.
3. Replace tag to <p> from <br>.
4. Move directory about GithubIusse component and Issues container.
5. Add / in link path.
6. Clean up import order.

lanieerts committed:
Modify additional issues about #48.
1. Combine filter function.
2. Apply ES6 templete literals.
3. Rename class name of GithubIssues -> GithubIssue
4. Delete unused div tag.
5. Delete unused actions & reducers.

lanieerts committed:
Add label, assignee, milstone filter.

Modify wrong letter.

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Merge branch 'master' into create-issues-page

lanieerts committed:
Add throttle event to text input.

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -45,6 +52,18 @@ class Issues extends React.Component {
+          <Debounce time="2000" handler="onKeyUp">
+            <input type="text" title="assignee"  placeholder="Assignee" list="assigneeList" onKeyUp={this.convertFilter} />
+          </Debounce>
+          <datalist id="assigneeList">
```
**`TODO`** : 이 리스트를 contributors 에서 가져옵시당

lanieerts replied:
action&reducer를 통해서 가져올까요?

pjhjohn replied:
넵 당연하군요..

pjhjohn commented:
Error

> 1. `hihi` 라고 침
> 2. `Fully` 라고 침
> 
> ```
> Object {milestone: "hihi"}
> xhr.js:158 GET https://api.github.com/repos/pjhjohn/jsonplaceholder-client/issues?milestone=hihi 422 (Unprocessable Entity)dispatchXhrRequest @ xhr.js:158xhrAdapter @ xhr.js:12
> Issues.js:220 Uncaught (in promise) TypeError: this.props.issues.map is not a function(…)render @ Issues.js:220_renderValidatedComponentWithoutOwnerOrContext @ ReactCompositeComponent.js:808_renderValidatedComponent @ ReactCompositeComponent.js:835_updateRenderedComponent @ ReactCompositeComponent.js:759_performComponentUpdate @ ReactCompositeComponent.js:739updateComponent @ ReactCompositeComponent.js:658receiveComponent @ ReactCompositeComponent.js:552receiveComponent @ ReactReconciler.js:126_updateRenderedComponent @ ReactCompositeComponent.js:761_performComponentUpdate @ ReactCompositeComponent.js:739updateComponent @ ReactCompositeComponent.js:658performUpdateIfNecessary @ ReactCompositeComponent.js:566performUpdateIfNecessary @ ReactReconciler.js:158runBatchedUpdates @ ReactUpdates.js:151perform @ Transaction.js:138perform @ Transaction.js:138perform @ ReactUpdates.js:90flushBatchedUpdates @ ReactUpdates.js:173closeAll @ Transaction.js:204perform @ Transaction.js:151batchedUpdates @ ReactDefaultBatchingStrategy.js:63enqueueUpdate @ ReactUpdates.js:201enqueueUpdate @ ReactUpdateQueue.js:25enqueueSetState @ ReactUpdateQueue.js:210ReactComponent.setState @ ReactComponent.js:64handleChange @ connect.js:301dispatch @ createStore.js:186dispatch @ instrument.js:460(anonymous function) @ index.js:28dispatch @ applyMiddleware.js:45action.payload.then._extends.payload @ index.js:25
> Issues.js:220 Uncaught TypeError: this.props.issues.map is not a function
> ```

pjhjohn commented:
https://developer.github.com/v3/#client-errors

pjhjohn commented:
milestone에 뭘 적어야할 지 확실치 않은데, 어느 경우든 위 에러는 안뜨는게 좋을 듯?
1. milestone을 검색한다 (by `id`)
2. milestone을 검색한다 (by `text`)
3. milestone에 등록된 issue를 검색한다 (by `id`)
4. milestone에 등록된 issue를 검색한다 (by `text`)

pjhjohn commented:
https://developer.github.com/v3/issues/#parameters-1

> If an integer is passed, it should refer to a milestone by its number field. If the string \* is passed, issues with any milestone are accepted. If the string none is passed, issues without milestones are returned.

3, 4인데, `none` 이 날라가면 마일스톤 없는걸 리턴하는군요

lanieerts committed:
Load contributors and apply autocomplete.

lanieerts committed:
Fix bug printing error about #49.

lanieerts committed:
Delete console.log function.

pjhjohn commented on src/actions/about.js:
```diff
@@ -1,9 +1,23 @@
-  const request = githubApi.get(`/issues`, query);
+  const request = githubApi.get(`/issues`, query)
+    .then((response) => {
+      if(!response.ok)
```
`response.ok` 는 타입이 어떻게 되나요? 200번대면 전부 ok인 건가.
그리고 ok가 아닐 때 `response.data`에는 무엇이 들어있나요? `undefined`인가요?

lanieerts replied:
`response.ok`의 타입은 `boolean`입니다. ok가 아닐때 `response.data`는 `undefined` 입니다.

pjhjohn replied:
`undefined`일 때 `[]`를 리턴하고 싶다면 다음과 같은 문법을 적용할 수 있습니다.

``` js
return response.data || [];
```

또한 JavaScript에는 다음과 같은 삼항연산자(Conditional Operator)가 존재합니다.
`condition` ? `conditioni_is_true` : `condition_is_false`

지금 `then` 에서 아무런 에러 처리를 해주고 있지 않은데, 만약 한다면 위에 설명한 것 들을 모두 적용하면 다음과 같은 코드가 될 겁니다.

``` js
return response.ok? response.data||[] : RETVAL_RESPONSE_NOT_OK;
```

**Tip** 삼항연산자는 **이런게 가능하다** 는 것이지 크게 권장하지는 않습니다. 보통 정말 간단하고 문맥도 너무 당연한데 코드가 불필요하게 길어지는 것을 막기 위한 `매크로` 정도의 느낌으로 사용합니다. 너무 긴 코드를 3항연산자를 사용해 표현하게 되면 쓸데없이 한 줄이 길어지거나 indent가 미관상, 가독성 상 저하되기 때문입니다.

lanieerts replied:
RETVAL_REPONSE_NOT_OK 는 action type을 return하기 위해서 쓰신건가요? 

>   const request = githubApi.get(`/issues`, query)
>     .then((response) => {
>       if(!response.ok)
>         response.data = [];
>       return response;
>     });

then 함수의 return 값이 request에 저장되는데 RETVAL_REPONSE_NOT_OK가 들어가도 되는건가요?

pjhjohn replied:
저건 예시로 든 것일 뿐이에요. 당연히 `RETVAL_RESPONSE_NOT_OK` 를 리턴하면 리퀘스트에 저장되겠죠. 지금은 ok가 아닐 때 핸들링을 하지 않을 뿐.. **action type을 return하기 위해 쓴 것이 아닙니다**

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -13,25 +14,36 @@ class Issues extends React.Component {
+  }
+
+  convertFilter = (filter) => {
+    let newFilter = { [filter.target.title]: filter.target.value };
```
`{ [filter.target.title]: filter.target.value };`부분에서 `[]`는 무슨 문법인가요..? Array가 아닌건가

lanieerts replied:
object element의 name값을 외부에 선언된 value값으로 지정하고 싶으면 저렇게 써야 한다고 버낟님이 찾아서 던져주셨습니다.

pjhjohn replied:
ok. got it.

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -13,25 +14,36 @@ class Issues extends React.Component {
+
+  convertFilter = (filter) => {
+    let newFilter = { [filter.target.title]: filter.target.value };
+    if(filter.target.value === "")
```
이것은 무조건 있어야 하는 아주 중요한 부분인가요?

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -13,25 +14,36 @@ class Issues extends React.Component {
+
+  convertFilter = (filter) => {
+    let newFilter = { [filter.target.title]: filter.target.value };
+    if(filter.target.value === "")
```
이 부분의 역할은 무엇인가요

lanieerts replied:
convertFilter 자체의 역할을 물어보시는 거면, input text로는 parameter name을 Dropdown button처럼 같이 묶어서 보낼 방법이 없는 것 같더라고요. 그래서 json parameter로 바꿔주는 역할을 합니다.
if문은 정보가 아예 없으면 string 값으로 none을 대신 집어넣습니다. 이게 parameter값에 아무것도 없으면("") 422 에러를 띄우더라고요.

lanieerts committed:
Remove unusable comment in Issues.js.

lanieerts committed:
Merge remote-tracking branch 'upstream/master'

lanieerts committed:
Merge branch 'master' into create-issues-page

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -13,25 +14,36 @@ class Issues extends React.Component {
+  };
 
   render() {
+    if(!this.props.contributors.length)
```
`{}` 를 붙이거나, inline으로 해주세요

lanieerts replied:
inline으로 해두겠습니다.

lanieerts committed:
Modify function about onKeyUp event.
1. Remove convertFilter function.
2. Input parameter directly from onKeyUp event to updateFilter function.
3. Simplificate using conditional opertator.

lanieerts committed:
Add delete statement when input value is "".

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -57,16 +49,18 @@ class Issues extends React.Component {
-            <input type="text" title="assignee"  placeholder="Assignee" list="assigneeList" onKeyUp={this.convertFilter} />
+          <Debounce time="1000" handler="onKeyUp">
+            <input type="text" title="assignee"  placeholder="Assignee" list="assigneeList"
+                   onKeyUp={(event) => this.updateFilter(event.target.value===""? {} : { [event.target.title]: event.target.value})} />
```
``` js
let _undefined = undefined;
let _null = null;
let _empty = "";
let _str = "str";
let test = (input) => {
  console.log(input);
  if(input) console.log("true");
  if(!input) console.log("false");
}
test(_undefined);
test(_null);
test(_empty);
test(_str);
```

이거 실행하면 `false` `false` `false` `true` 뜨는걸 봐서는

``` js
event.target.value===""? {} : { [event.target.title]: event.target.value}
```

를

``` js
event.target.value? { [event.target.title]: event.target.value} : {}
```

로 해도 되지 않을까

pjhjohn commented on src/containers/About/Issues.js:
```diff
@@ -20,6 +20,8 @@ class Issues extends React.Component {
   updateFilter = (filter) => {
     const newFilter = Object.assign(this.state.filter, filter);
+    for(let value in newFilter)
+      if(newFilter[value] === "") delete newFilter[value];
```
`filter`의 `key`값이 곧 지울 타겟인데, 굳이 전체를 조사할 필요가 있나요?

lanieerts committed:
Modify parameter to updateFilter and clean up code.
